test/system/src/test/java/io/pravega/test/system/MultiReaderTxnWriterWithFailoverTest.java

```diff
@@ -0,0 +1,496 @@
+
+                    //each transaction has 50 events
+                    for (int j = 0; j < 50; j++) {
+                        long value = data.incrementAndGet();
+                        Thread.sleep(100);
+                        transaction.writeEvent(String.valueOf(value), value);
+                        log.debug("Writing event: {} into transaction: {}", value, transaction);
+                    }
+                    //commit Txn
+                    transaction.commit();
+
+                    //wait for transaction to get committed
+                    checkTxnStatus(transaction, eventWriteCount).get();
+                } catch (Throwable e) {
+                    log.warn("Exception while writing events in the transaction: {}", e);
+                    log.debug("Transaction with id: {}  failed", transaction.getTxnId());
+                }
+            }
+        }, executorService);
+    }
+
+    private CompletableFuture<Void> checkTxnStatus(Transaction<Long> txn,
+                                                   final AtomicLong eventWriteCount) {
+        return CompletableFuture.runAsync(() -> {
+            Transaction.Status status = retry.retryingOn(MultiReaderTxnWriterWithFailoverTest.TxnNotCompleteException.class)
+                    .throwingOn(RuntimeException.class).run(() -> txn.checkStatus());
+            log.debug("Transaction: {} status: {}", txn, txn.checkStatus());
+            if (status.equals(Transaction.Status.COMMITTED)) {
+                eventWriteCount.getAndIncrement();
+                log.info("Event write count: {}", eventWriteCount.get());
+            } else if (status.equals(Transaction.Status.ABORTED)) {
+                log.debug("Transaction with id: {} aborted", txn.getTxnId());
+            } else {
+                throw new TxnNotCompleteException();
+            }
+        });
+    }
+
+    private Transaction<Long> createTransaction(EventStreamWriter<Long> writer, final AtomicBoolean exitFlag) {
+        Transaction<Long> txn = null;
+        try {
+            //Default max scale grace period is 30000
+            txn = writer.beginTxn(5000, 3600000, 29000);
+            log.info("Transaction created with id:{} ", txn.getTxnId());
+        } catch (RuntimeException ex) {
+            log.info("Exception encountered while trying to begin Transaction ", ex.getCause());
+            if (ex instanceof io.grpc.StatusRuntimeException && !exitFlag.get()) {
+                //Exit flag is true no need to retry.
+                log.warn("Cause for failure is {} and we need to retry", ex.getClass().getName());
+                throw new TxnCreationFailedException(); // we can retry on this exception.
+            } else {
+                throw ex;
+            }
+        }
+        return txn;
+    }
+
+    private CompletableFuture<Void> startReading(final ConcurrentLinkedQueue<Long> readResult, final AtomicLong writeCount, final
```

Comment:
Here and above: Rather than passing mutable members, utilize the return value for results and member variables for things like the booleans/longs that are shared.